File: 84-largest-rectangle-in-histogram/largest-rectangle-in-histogram.cpp
class Solution {
public:
    int largestRectangleArea(vector<int>& heights) {
        heights.push_back(0);
        int ans=0;
        stack<int> s;
        for(int i=0;i<heights.size();i++)
        {
            while(!s.empty() && heights[s.top()]>heights[i])
            {
                int height=heights[s.top()];s.pop();
                int width;
                if(!s.empty())
                {
                    width=i-s.top()-1;
                }
                else{
                    width=i;
                }
                ans=max(ans,height*width);
            }
            s.push(i);
        }

        return ans;
    }
};

File: 84-largest-rectangle-in-histogram/largest-rectangle-in-histogram.java
class Solution {
    public int largestRectangleArea(int[] heights) {
        int n = heights.length;
        Stack<int[]> st = new Stack<>();
        int ans=0;  
        for(int i=0;i<n;i++)
        {
            int ind=i;
            while(!st.isEmpty() && st.peek()[1]>heights[i])
            {
                int[] x = st.pop();
                ind=Math.min(ind,x[0]);
                ans=Math.max(ans,x[1]*(i-x[0]));
                ans=Math.max(ans,x[1]);
            }
            st.push(new int[]{ind,heights[i]});
        }
        while(!st.isEmpty())
        {
            int[] x=st.pop();
            ans=Math.max(ans,x[1]*(n-x[0]));
        }
        return ans;
    }
}